### PR TITLE
feat: adds app switcher lab component

### PR DESCRIPTION
**Pull request by michaeltamaki-okta**
[OKTA-825118](https://oktainc.atlassian.net/browse/OKTA-825118)

## Summary

feat: adds app switcher lab component

Building out the presentation layer of the "App Switcher" https://odyssey.okta.design/latest/components/app-switcher/design-mla5Wn5G (now called "Nav Rail").
Figma: https://www.figma.com/design/zh7A9gjaDlI50Hctdvd1OB/Odyssey-Components?node-id=22141-142254&node-type=frame&t=Qc7MjiWUFNtOh9Vd-0
Playground: https://6485e4e33982ca4977fff8d81ef98f680e713a9c.ods.dev/?path=/docs/labs-components-appswitcher--docs

### Code structure

- [x] Organize code using a modular component architecture.
- [x] Use semantic HTML for clear and meaningful structure.
- [x] Break down complex components into smaller, reusable parts.

### Accessibility implementation

- [x] Apply ARIA attributes to indicate roles and relationships.
- [x] Test focus management and ensure interactive elements are keyboard accessible.
- [x] Provide visual alternatives for auditory elements.

### Internationalization (i18n) implementation

- [x] Separate any text content from the component's code for translation.
- [x] Test RTL support in Storybook

Screenshot 2024-11-18 at 8 44 24 AM

### Cross-browser compatibility

- [x] Ensure consistent behavior and appearance across browsers.

#### Chrome 
Screenshot 2024-11-15 at 10 14 34 AM
Screenshot 2024-11-15 at 10 14 45 AM

#### Safari
Screenshot 2024-11-15 at 10 15 13 AM
Screenshot 2024-11-15 at 10 15 21 AM

#### Firefox
Screenshot 2024-11-15 at 10 17 37 AM
Screenshot 2024-11-15 at 10 17 43 AM

### Documentation

- [x] Include code snippets, demos, and examples for various use cases.
- [x] Provide guidance on customization and handling special cases.

### Finalize code

- [x] Interaction designers, UI eng and visual designers need to review that the Storybook code matches the intended Figma design
- [ ] Interaction designers need to confirm Storybook code matches Supernova
- [ ] Interaction and visual designers merge Figma and code at the same time
- [ ] Share updates to the [#design-system-team](https://okta.slack.com/archives/C02JL12CJPQ)

### QA

- [x] Include necessary unit tests to continuously verify component functionality
- [x] Screen reader test error/success states or message
- [x] Test edge cases and error scenarios to validate robustness

### Testing and validation

- [ ] Share work in [Craft](https://docs.google.com/document/d/1dqaEzKW3SlKccw2ILjX7y3LQY0toeNe9OuwpOs2XxcI/edit#heading=h.j8zakyclqze5) to get feedback from the Odyssey team
- [ ] Share work to get feedback from pillar design captains or product designers in [#design-team](https://okta.slack.com/archives/G7REPJFUY) as customer feedback is essential to help us validate our assumptions, identify pain points, and improve our user experience

### CODE REVIEW

File: packages/odyssey-react-mui/src/labs/NavRail/OktaAura.tsx [hunk 1]
```diff
@@ -10,14 +10,20 @@
  * See the License for the specific language governing permissions and limitations under the License.
  */
 
+import styled from "@emotion/styled";
 import { memo } from "react";
 
+const OktaAuraSvgComponent = styled("svg")(() => ({
```
Comment (michaeltamaki-okta) [inline packages/odyssey-react-mui/src/labs/NavRail/OktaAura.tsx:16]: Copied over from the other PR
@bryancunningham-okta https://github.com/okta/odyssey/pull/2415#discussion_r1844484001
> @michaeltamaki-okta I think we can use the svg that currently lives in the `SideNav` folder. We can just wrap it in a styled component and apply these sizes to it. Rather than duplicate the SVG with a different size.

The SVG in the `SideNav` folder isn't being used (at least as far as I can tell), so I actually moved it to the `NavRail` folder here.

File: packages/odyssey-react-mui/src/labs/NavRail/NavRailApp.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    borderRadius: odysseyDesignTokens.BorderRadiusTight,
+    transition: `background-color ${odysseyDesignTokens.TransitionDurationMain}`,
+
+    "&:hover, &:focus": {
+      backgroundColor: odysseyDesignTokens.HueNeutral50,
+    },
+  }),
+);
+
+const NavRailAppImgComponent = styled("img")(() => ({
```
Comment (michaeltamaki-okta) [inline packages/odyssey-react-mui/src/labs/NavRail/NavRailApp.tsx:61]: Copied from the other PR @bryancunningham-okta https://github.com/okta/odyssey/pull/2415#discussion_r1844488328

> IMHO, It would be great if we could render the SVGs inline, rather than using images here

The plan is to load the SVGs from monolith/CDN (https://github.com/atko-eng/okta-ui/pull/8012 https://github.com/atko-eng/okta-core/pull/103856). The pros of this approach is that any app icon changes/additions/removals don't require an Odyssey version bump since it all comes from the backend. The downside is that we can't do inline SVGs like you suggested.

File: packages/odyssey-react-mui/src/labs/NavRail/NavRailApp.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      return (
+        <NavRailAppLinkComponent
+          {...muiProps}
+          odysseyDesignTokens={odysseyDesignTokens}
+          isSelected={isSelected}
+          href={linkUrl}
+        >
+          <NavRailAppImgComponent
+            src={isSelected ? appIconSelectedUrl : appIconDefaultUrl}
+            alt="" // Tell screen reader to ignore the image; the Tooltip describes the link
```
Comment (michaeltamaki-okta) [inline packages/odyssey-react-mui/src/labs/NavRail/NavRailApp.tsx:98]: role=presentation

File: packages/odyssey-react-mui/src/labs/NavRail/NavRail.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import {
+  DesignTokens,
+  useOdysseyDesignTokens,
+} from "../../OdysseyDesignTokensContext";
+import { OktaAura } from "./OktaAura";
+import { NavRailApp, NavRailAppProps } from "./NavRailApp";
+
+const NAV_RAIL_WIDTH = "64px";
+
+export type NavRailProps = {
```
Comment (michaeltamaki-okta) [inline packages/odyssey-react-mui/src/labs/NavRail/NavRail.tsx:24]: Add `visibilityState = invisible | loading | loaded`

### How should we handle the loading state?

Skeleton
- Enduser: Skeleton shows, then disappears (pop in during load)
- Admin: Skeleton shows, then gets populated

No Skeleton ✅ 
- Enduser: Skeleton does not show, then disappears
- Admin: Skeleton does not show, then nav rail pops in
- Could we mitigate this via storing in local storage to prevent pop in? Would pop in the first load, but would not pop in after that.
- Also consider that this is just for the MVP where it is only shown for the admin. In the future, when it is shown to all users, we could switch to skeleton.

Full screen loading screen (prevents pop in)
- Would block showing the app until we get the data. Unacceptable to downstream teams.

Keep side nav the same width regardless of app switcher
- No, would break design assumptions

File: packages/odyssey-storybook/src/components/odyssey-labs/NavRail/NavRail.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    return (
+      <div style={{ height: "100vh" }}>
+        <NavRail {...props} />
+      </div>
+    );
+  },
+  play: async ({ canvasElement, step }: PlaywrightProps<NavRailProps>) => {
+    const canvas = within(canvasElement);
+
+    await step("Nav rail successfully loads", async () => {
```
Comment (michaeltamaki-okta) [inline packages/odyssey-storybook/src/components/odyssey-labs/NavRail/NavRail.stories.tsx:93]: shows all items

File: packages/odyssey-storybook/src/components/odyssey-labs/NavRail/NavRail.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    return (
+      <div style={{ height: "100vh" }}>
+        <NavRail {...props} />
+      </div>
+    );
+  },
+  play: async ({ canvasElement, step }: PlaywrightProps<NavRailProps>) => {
+    const canvas = within(canvasElement);
+
+    await step("Nav rail does not load", async () => {
```
Comment (michaeltamaki-okta) [inline packages/odyssey-storybook/src/components/odyssey-labs/NavRail/NavRail.stories.tsx:114]: hides when no items

**Review by michaeltamaki-okta — COMMENTED** [review]
Added notes from meeting with @KevinGhadyani-Okta

File: packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+} from "../../OdysseyDesignTokensContext";
+import { OktaAura } from "./OktaAura";
+import {
+  AppSwitcherApp,
+  AppSwitcherAppIconData,
+  AppSwitcherAppSkeleton,
+} from "./AppSwitcherApp";
+import { TOP_NAV_HEIGHT } from "../TopNav";
+
+const APP_SWITCHER_WIDTH = `${64 / 14}rem`;
```
Comment (bryancunningham-okta) [inline packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx:27]: @michaeltamaki-okta Is this value meant to scale with the font size? Or, should it be a static px value?

File: packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+const AppSwitcherOktaAuraWrapperComponent = styled("div", {
+  shouldForwardProp: (prop) => prop !== "odysseyDesignTokens",
+})(({ odysseyDesignTokens }: { odysseyDesignTokens: DesignTokens }) => ({
+  width: "100%",
+  height: TOP_NAV_HEIGHT,
+  display: "flex",
+  alignItems: "center",
+  justifyContent: "center",
+  marginBottom: odysseyDesignTokens.Spacing4,
```
Comment (bryancunningham-okta) [inline packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx:56]:
```suggestion
  marginBlockEnd: odysseyDesignTokens.Spacing4,
```

File: packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcherApp.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+import {
+  DesignTokens,
+  useOdysseyDesignTokens,
+} from "../../OdysseyDesignTokensContext";
+import { Tooltip } from "../../Tooltip";
+import { MuiPropsContext, MuiPropsContextType } from "../../MuiPropsContext";
+
+const APP_SIDE_LENGTH_VAL = 36;
+const APP_SIDE_LENGTH_REM = `${APP_SIDE_LENGTH_VAL / 14}rem`;
+const APP_ICON_SIDE_LENGTH = `${32 / 14}rem`;
```
Comment (bryancunningham-okta) [inline packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcherApp.tsx:25]: We should probably be using token values for these. If we don't have token values for these sizes already, the sizes are likely incorrect. We'd have to go back to design and tell them we're using a token that already exists

**Review by benschell-okta — DISMISSED** [review]
I do not feel qualified to review the core component, but the changes UiShell looks exactly like I hoped/expected (and that's primarily what my approval relates to)! Great work!

File: packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx (hunk 7 above)
Comment (bryancunningham-okta) [reply packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx:27]: If it should be a rem value, we should probably use one of our spacing tokens here

Comment (KevinGhadyani-Okta) [reply packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcher.tsx:27]: That would be Spacing9 I think. The same as top nav.

File: packages/odyssey-storybook/src/components/odyssey-labs/AppSwitcher/AppSwitcher.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    );
+  },
+  play: async ({ canvasElement, step }: PlaywrightProps<AppSwitcherProps>) => {
+    const canvas = within(canvasElement);
+
+    await step("App switcher shows no items, but is loaded", async () => {
+      await expect(canvas.getAllByRole("navigation")).toHaveLength(1);
+      await expect(canvas.queryAllByRole("listitem")).toHaveLength(0);
+      await expect(canvas.queryAllByRole("link")).toHaveLength(0);
+    });
```
Comment (KevinGhadyani-Okta) [inline packages/odyssey-storybook/src/components/odyssey-labs/AppSwitcher/AppSwitcher.stories.tsx:119]: You can check on the navigation "to be visible". I wonder if navigations should have a label because someone's not gonna know what they're selecting. I think there's a difference between primary and secondary navigations in ARIA. Something you'll wanna check into.

**Review by KevinGhadyani-Okta — COMMENTED** [review]
Looks good! I made a few small comments.

File: packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcherApp.tsx [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+export const AppSwitcherApp = ({
+  appIconDefaultUrl,
+  appIconSelectedUrl,
+  appName,
+  label,
+  linkUrl,
+  selectedAppName,
+}: AppSwitcherAppProps) => {
+  const odysseyDesignTokens = useOdysseyDesignTokens();
+  const isSelected = appName === selectedAppName;
```
[the remark this answers is not in the document]
Comment (michaeltamaki-okta) [reply packages/odyssey-react-mui/src/labs/AppSwitcher/AppSwitcherApp.tsx:95]: We'd have to include this as part of `renderUiShell`

**Review by bryancunningham-okta — DISMISSED** [review]

**Review by KevinGhadyani-Okta — APPROVED** [review]
